### ISSUE TITLE
Hyperion dust grid is rotated with respect to star particles with Enzo data

### ISSUE THREAD

**Issue by anchwr** [dnarayanan/powderday#113]
There appears to be an issue with the yt-hyperion interface that results in the hyperion dust grid being on a different set of axes than the star particles for Enzo data. Imaging the sample Enzo galaxy, for instance, reveals that the dust lanes are at a 90 degree angle to the stellar disk. I think this is happening when from_yt is called in the enzo tributary. Looking at the [source code for this function](https://github.com/hyperion-rt/hyperion/blob/2159b39f5cafaeb653f2541502eef194a78102d6/hyperion/grid/yt3_wrappers.py#L199), it looks like hyperion is expecting to read in data with zyx ordering (presumably because Fortran is column-major?), but the data that it's being fed is actually xyz ordered. 

I've been able to compensate for this with the sample Enzo galaxy by adding a 90 degree rotation of the star particles about the y-axis (i.e., if a particle's position was (x0,y0,z0) the new position is (-z0,y0,x0)) to stars_coordinate_boost in grid_construction.py after the star particles are re-centered. This results in the dust lanes appearing in the correct plane and tracking the spiral arms of the galaxy. However, the same solution doesn't work for the FOGGIE simulations - even after rotating the star particles, the dust lanes and the stellar disks of the galaxies are still in different planes - so this must be a little more complicated than just a single rotation. Additionally, rotating the star particles is not a great solution, since this changes the orientation of a galaxy relative to what it was in the original simulation and therefore takes away the user's intuition for what they'll see when they look down a given axis. I'm hoping both of these problems might be solved if there's a way to feed from_yt data in the format that it expects, but I haven't found a way to do this using yt. Based on [this](https://github.com/yt-project/yt/issues/893) it looks like this might already be a known issue with octree data, but I'm hoping we can find a fix.

**Comment by dnarayanan** [comment]
thanks for reporting this (and for the sleuthing!) and I'm sorry that this is an issue.   I have to admit we never checked the images against projection plots when putting the Enzo front end together so this didn't flag up as an error.

to be 100% sure I understand -- you think that the AMR grid is being read in correctly from yt in Hyperion, but that the stars are being dumped out with coordinates xyz, when Hyperion expects zyx?

**Comment by dnarayanan** [comment]
I'll loop in @jwise77 and @snigdaa as well in case they have any insight or have seen anything like this in their models.

**Comment by anchwr** [comment]
I guess it depends on whether you think the orientation of the original simulation is correct or the way in which hyperion reorients the AMR grid is correct. The problem is that the star particle positions reflect the simulation's orientation, but this doesn't match the hyperion AMR grid's orientation, so when the star particles are placed onto the hyperion AMR grid as sources, their positions relative to the dust are incorrect. As far as I can tell, hyperion keeps the x,y,z ordering for source positions (presumably because each individual position is a 1D array), but the hyperion AMR grid gets reoriented because it's a multidimensional array (so the difference between row-major and column-major ordering matters). I guess my opinion would be that we should try to keep the orientation of the original simulation and reorient the hyperion AMR grid.

**Comment by dnarayanan** [comment]
Ah okay I see.  So if I can summarize what you've already said, it sounds like there's 2 options:

[1] **Change the ordering of how Hyperion creates the grid from zyx to xyz to match the xyz ordering of how yt lists the star particles.**. This looks by eye like it would be involved around[ here]([url](https://github.com/hyperion-rt/hyperion/blob/2159b39f5cafaeb653f2541502eef194a78102d6/hyperion/grid/yt3_wrappers.py#L292)).  This has the advantage of then creating hyperion/powderday grids that are ordered similarly to the simulation, and therefore preserves intuition when going from simulation maps to observational ones.

Functionally, we'd want to maintain our own fork of Hyperion if we did this thats exclusive for powderday, but thats not a big deal.  The larger issue is that I don't really know how Enzo-type grids work (or how they're stored in yt).  I suspect something akin to the [find_orde](https://github.com/hyperion-rt/hyperion/blob/2159b39f5cafaeb653f2541502eef194a78102d6/hyperion/grid/yt3_wrappers.py#L118)r octree-reordering would have to be enabled.   I'd have to lean on you (or someone with some Enzo knowledge) to enable this.

[2] **Change the ordering of the star particles from xyz-->zyx in post-processing**. This has the advantage of being the simplest solution, and is implementable in powderday alone (i.e., no need for updates to hyperion), though has two disadvantages: 

(a) the maps would be oriented differently than the yt maps (this is already the situation for octree based grids as @mabruzzo figured out a while ago)

(b) while this seems to work for some simulations, it doesn't for others (the FOGGIE one).


I can look into [2] and play around with it (I wonder if for some reason putting the re-ordering either in the Enzo front end or tributary might be useful instead of the coordinate boost; I'd need to look/think about it).   Is the FOGGIE simulation in that google drive link you sent me in the Fall?

**Comment by anchwr** [comment]
If I remember right, the simulation I uploaded is a slightly lower resolution version of the simulation I've been running powderday on. The lower resolution simulations tend to produce fluffier galaxies, so it might be a little bit harder to tell where the dust lanes should be, but I think that one should be okay. 

Figuring out what transformation is required to get the stellar disk on the same axes as the hyperion AMR grid definitely seems like the easier solution. I've experimented with a few different rotations and haven't found anything that works for both the sample Enzo galaxy and the FOGGIE galaxy, but it's definitely possible that I'm thinking too simplistically about it. My hope was that I could figure that out and then figure out how to translate the changes to the Enzo tributary so that there wouldn't need to be an "if Enzo, then..." statement in the coordinate boost function, but maybe you're right and I should have started from the tributary.

I can see if I can figure out how something like find_order would work for Enzo data. I'm still fairly new to grid codes, but I can hopefully get some help from the more knowledgeable people on the FOGGIE team :) I could also try posting to the yt slack, if you think that would be helpful. Being able to specify zyx ordering when we save the dataset prior to passing it to hyperion would be ideal, since that could be implemented purely within enzo_zoom in powderday and would hopefully solve all of the reorientation issues. Someone asked a somewhat related question on the yt slack a few days ago and the answer makes me think that this functionality doesn't exist, but it seems like it could be worth asking about.

**Comment by dnarayanan** [comment]
thanks for checking on yt slack!  I saw your message there (though didn't see the other related message).    

I'll play around with option 2 in the mean time and see if I can get anywhere.  please do update if you get anywhere on this as well.

**Comment by dnarayanan** [comment]
actually I have a quick question - I think I don't quite know how to load the snapshot correctly.  I have:

``` 
[desika.narayanan@c23b-s38 RD0042]$ ls
RD0042          RD0042.cpu0027  RD0042.cpu0062  RD0042.cpu0108  RD0042.cpu0164  RD0042.cpu0191  RD0042.cpu0214  RD0042.cpu0263  RD0042.cpu0317  RD0042.cpu0341  RD0042.cpu0409  RD0042.cpu0451  RD0042.cpu0468  RD0042.cpu0498
RD0042.cpu0000  RD0042.cpu0028  RD0042.cpu0068  RD0042.cpu0122  RD0042.cpu0166  RD0042.cpu0195  RD0042.cpu0225  RD0042.cpu0265  RD0042.cpu0318  RD0042.cpu0350  RD0042.cpu0412  RD0042.cpu0452  RD0042.cpu0469  RD0042.cpu0503
RD0042.cpu0001  RD0042.cpu0042  RD0042.cpu0069  RD0042.cpu0123  RD0042.cpu0176  RD0042.cpu0199  RD0042.cpu0227  RD0042.cpu0278  RD0042.cpu0324  RD0042.cpu0357  RD0042.cpu0413  RD0042.cpu0454  RD0042.cpu0472  RD0042.cpu0504
RD0042.cpu0010  RD0042.cpu0043  RD0042.cpu0082  RD0042.cpu0124  RD0042.cpu0181  RD0042.cpu0200  RD0042.cpu0228  RD0042.cpu0284  RD0042.cpu0330  RD0042.cpu0362  RD0042.cpu0437  RD0042.cpu0456  RD0042.cpu0475
RD0042.cpu0018  RD0042.cpu0046  RD0042.cpu0091  RD0042.cpu0135  RD0042.cpu0182  RD0042.cpu0201  RD0042.cpu0233  RD0042.cpu0293  RD0042.cpu0334  RD0042.cpu0376  RD0042.cpu0442  RD0042.cpu0457  RD0042.cpu0476
RD0042.cpu0019  RD0042.cpu0048  RD0042.cpu0098  RD0042.cpu0158  RD0042.cpu0185  RD0042.cpu0205  RD0042.cpu0237  RD0042.cpu0312  RD0042.cpu0335  RD0042.cpu0377  RD0042.cpu0445  RD0042.cpu0459  RD0042.cpu0478
RD0042.cpu0022  RD0042.cpu0050  RD0042.cpu0100  RD0042.cpu0163  RD0042.cpu0186  RD0042.cpu0206  RD0042.cpu0243  RD0042.cpu0315  RD0042.cpu0337  RD0042.cpu0390  RD0042.cpu0450  RD0042.cpu0463  RD0042.cpu0496```

but then this fails:

```In [2]: ds = yt.load('RD0042')
yt : [ERROR    ] 2021-01-29 07:54:29,368 Couldn't figure out output type for RD0042
---------------------------------------------------------------------------
YTOutputNotIdentified                     Traceback (most recent call last)
<ipython-input-2-67453e365f17> in <module>
----> 1 ds = yt.load('RD0042')

~/miniconda3/envs/pd4env/lib/python3.6/site-packages/yt-4.0.dev0-py3.6-linux-x86_64.egg/yt/convenience.py in load(*args, **kwargs)
     80                 return output_type_registry[n](fn)
     81         mylog.error("Couldn't figure out output type for %s", args[0])
---> 82         raise YTOutputNotIdentified(args, kwargs)
     83
     84     mylog.error("Multiple output type candidates for %s:", args[0])

YTOutputNotIdentified: Supplied ('RD0042',) {}, but could not load!```

**Comment by anchwr** [comment]
Well that's bizarre. It looks like a bunch of the outputs are missing. That must have somehow happened when I made the tar file. Let me make a new one (and check it this time...) and upload it. Sorry about that!

**Comment by anchwr** [comment]
It just finished uploading. Let me know if you don't get the share link! 
I messaged the rest of the FOGGIE team about the yt-hyperion interface issue and Brian O'Shea pointed out that hyperion is transposing the grid, rather than simply rotating it, which is likely why my attempts to relocate the star particles weren't terribly successful.

**Comment by dnarayanan** [comment]
I did get this -thank you!   I'm checking out a bit early (now) to take my kids camping but will download this on Monday.   but, this may be a bit more complicated than we think like you mentioned.  that's good advice by Brian.

perhaps trying to mimic what Hyperion is doing with the grid on the star particles will be a good place to start.

**Comment by anchwr** [comment]
I've found a fix for this, although I don't know that it's the ideal thing to do. I wasn't able to find a way to transpose the grid as a yt object, so I moved hyperion's from_yt/yt_dataset_to_amr_grid function into the enzo tributary and changed the data as it was being read into the hyperion AMR grid within that function. The runs now produce dust lanes in the right locations for both EnzoDisk and the FOGGIE galaxies (I've put a couple of example images below). It would be great if this set-up could just be an option in hyperion (similar to what's done with octree grids), but having it be part of the enzo tributary doesn't impact the structure or speed of the pipeline (as far as I can tell). 
![Tempest Famr hidust PAH 1e9](https://user-images.githubusercontent.com/14131084/111354029-42c12700-865c-11eb-86bf-771ccb387fb4.png)
![EnzoDisk Famr hidust PAH 1e6](https://user-images.githubusercontent.com/14131084/111354071-4d7bbc00-865c-11eb-9594-aa9bdaeba841.png)
